### PR TITLE
PCA plot

### LINKED ISSUE #96: We'd like to implement PCA decomposition plot
Principal component analysis (PCA) to decompose high dimensional data into two or three dimensions so that each instance can be plotted in a scatter plot. 

As part of this issue, we'll need to research this plot and implement it as part of the package.
We should test it and add it to the CI, in addition to documenting the code and how to use it via examples.
Below is how it should look like.

### PR DESCRIPTION
**Pull request by neelasha23**
## Describe your changes
PCA Decomposition plot

## Issue ticket number and link
Closes #96

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] I have added thorough tests (when necessary).
- [x] I have added the right documentation (when needed). Product update? If yes, write one line about this update.

----
:books: Documentation preview :books:: https://sklearn-evaluation--233.org.readthedocs.build/en/233/

### CODE REVIEW

**Comment by neelasha23** [comment]
Need a suggestion here. Can we show all the principal components plots as in [here](https://plotly.com/python/pca-visualization/) (`Visualize all the principal components` section), or will it be too many?
@edublancas

**Comment by idomic** [comment]
@neelasha23 besides the above, is it ready for review?

**Comment by neelasha23** [comment]
yes it's ready. 
I wanted to add some points on interpreting the principal component plot in the user guide when I found this link. Maybe it would be helpful , but wasn't sure.
Other than that it's ready

**Comment by idomic** [comment]
I think it might be helpful, but what happens if you have 8 - 9 dimensions? It can become really busy.

**Comment by edublancas** [comment]
looks like the `n_components` isn't used currently? putting 2 and 3 gives the same results

image

If so, let's remove it. Regarding your point of visualizing all principal components, I think it makes sense; let's leave it for v2. I think we can re-add `n_components` in v2. so 2 components will generate the current graph but >2 can generate a grid of all pairwise comparisons.

**Comment by edublancas** [comment]
also, let's rename the second argument Y to y, to be consistent with other plots

File: src/sklearn_evaluation/plot/pca.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    else:
+        Y = Y if isinstance(Y, np.ndarray) else np.array(Y)
+
+        if len(Y.shape) < 2:
+            Y = np.expand_dims(Y, 1)
+        new_principalComponents = np.hstack((principalComponents, Y))
+
+        from random import randint
+        targets = np.unique(new_principalComponents[..., -1])
+        colors = ['#%06X' % randint(0, 0xFFFFFF)
```
Comment (yafimvo) [inline src/sklearn_evaluation/plot/pca.py:88]: I think we can export this logic to a helper function and reuse it later in other plots (to make coloring and styling consistent). something like `get_default_colors(targets)` (we don't need to export it yet, but refactoring will be easier in the feature).

File: src/sklearn_evaluation/plot/pca.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+@modify_exceptions
+@SKLearnEvaluationLogger.log(feature='plot')
+def pca(X, Y=None, target_names=None, n_components=2, colors=None, ax=None):
+    """
+    Plot principle component analysis curve.
+
+    Parameters
+    ----------
+
+    X : array-like, shape = [n_samples, n_features]
```
Comment (yafimvo) [inline src/sklearn_evaluation/plot/pca.py:28]: description for the "colors" parameter is missing

File: tests/test_pca.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+@image_comparison(baseline_images=["pca"], extensions=["png"], remove_text=False)
+def test_pca():
+    pca(X, y, colors=["r", "b"])
+
+
+@image_comparison(
+    baseline_images=["pca_without_y"], extensions=["png"], remove_text=False
+)
+def test_pca_without_y():
```
Comment (yafimvo) [inline tests/test_pca.py:22]: we can add another test:  `test_pca_without_colors`

**Review by yafimvo — CHANGES_REQUESTED** [review]
lgtm.
added minor changes

**Comment by edublancas** [comment]
@neelasha23: please don't edit comments since it makes the discussion difficult to read (I noticed you added answers in my comment).

Pasting the original points (with some edits) of my earlier comment

---

looks like the n_components isn't used currently? putting 2 and 3 gives the same results. Even if the argument is passed to `PCA`, it has no use because passing 2,3,4,5... won't have any effect since we're plotting the first two components anyway. We should remove this argument, unless we use it for determining which components to plot. see my comment below

> Maybe we can take another parameter `plot_components` which can have following values:
> 
>     1. `None` (default to first component vs second component)
> 
>     2. A list of pairs [(1,2), (1,3)] : so this will generate two plots , first component vs second component and first component vs third component
> 
>     3. `All` : generate all possible pairs of components


It's hardly the case that you want to plot specific components, the common case is to plot the first N, since they contain the most information (and information gain decreases as we add more components), so I suggest using `n_components` for this:

1. by default it should be 2
2. if 1, throw an error
3. if >2, plot all pairwise comparisons

**Comment by neelasha23** [comment]
Yes I meant to quote the reply, I think I clicked on `Edit` by mistake.

**Comment by yafimvo** [comment]
@neelasha23 there are conflicts

File: src/sklearn_evaluation/plot/pca.py (hunk 1 above)
Comment (neelasha23) [reply src/sklearn_evaluation/plot/pca.py:88]: Agreed. But I think we can open an issue on this for now and work on it once more elaborate specifications are listed down (since it should work for all plots)

File: tests/test_pca.py (hunk 3 above)
Comment (neelasha23) [reply tests/test_pca.py:22]: Added but since colors are randomly generated in every run the `image_comparison` hasn't been added

**Comment by idomic** [comment]
@neelasha23 please rebase on main

**Comment by neelasha23** [comment]
This is ready for review.
Docs build has passed. CI is taking long time but I think it would pass.
@edublancas @idomic

File: docs/dimensionality_reduction/pca.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+```{code-cell} ipython3
+from sklearn.neighbors import KNeighborsClassifier
+X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state=42) 
+knn = KNeighborsClassifier(n_neighbors=7)
+knn.fit(X_train, y_train) 
+print("Accuracy before PCA:", knn.score(X_test, y_test)) 
+```
+
+Now let's apply PCA on the data and retrain a model. We can see that the accuracy after PCA is as good as without PCA.
```
Comment (idomic) [inline docs/dimensionality_reduction/pca.md:75]: I would add here, ...as good as without PCA, meaning we can get the same results with X (faster, less data etc).

File: docs/_toc.yml [hunk 5]
```diff
@@ -36,6 +36,10 @@ parts:
     - file: optimization/grid_search
     - file: optimization/feature_ranking
 
+  - caption: Dimensionality Reduction
```
Comment (idomic) [inline docs/_toc.yml:39]: We might want to unify this topic under use cases. Thoughts?

File: src/sklearn_evaluation/plot/pca.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    if target_names and colors and len(target_names) != len(colors):
+        raise ValueError(
+            f"Length received. target_names : {len(target_names)}, "
+            f"colors : {len(colors)}. Lengths should be same"
+        )
+
+    if ax is not None:
+        ax_count = int((n_components * (n_components - 1)) / 2)
+        if len(ax) != ax_count:
+            raise ValueError(f"Number of axes passed should be {ax_count}")
```
Comment (idomic) [inline src/sklearn_evaluation/plot/pca.py:35]: Are those value errors printing the slack community link?

File: tests/test_pca.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    assert ax is out_ax[0]
+
+
+def test_nan_error(ploomber_value_error_message):
+    with pytest.raises(ValueError, match=ploomber_value_error_message):
+        pca(np.array([[1, float("nan"), 2], [3, 0, 6]]), np.array([1, 2]))
+
+
+def test_component_error(ploomber_value_error_message):
+    with pytest.raises(ValueError, match=ploomber_value_error_message):
```
Comment (idomic) [inline tests/test_pca.py:75]: I see you're testing for the community slack message, ignore last comment.

**Review by idomic — CHANGES_REQUESTED** [review]
Looks good!

I think we also need to add telemetry on this plot and test for it's mocked call.
+ I didn't see a changelog entry.

File: src/sklearn_evaluation/plot/pca.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+        n_features is the number of features
+
+    y : array-like or list or None
+        set None if ignored otherwise, pass the targets here
+
+    target_names: list, optional
+        list of target variable names
+
+    n_components : int, float or 'mle', default=None
+        Number of components to keep.
```
Comment (edublancas) [inline src/sklearn_evaluation/plot/pca.py:58]: the default needs updates here, it says None but it's two. also let's add more detail: "if 2, it plots the first and second components, it 3 or more, it plots all pairwise comparisons"

**Review by edublancas — CHANGES_REQUESTED** [review]

File: docs/_toc.yml (hunk 5 above)
Comment (neelasha23) [reply docs/_toc.yml:39]: yes I thought so too, but wasn't sure which one. maybe clustering? or create another topic named `Feature Selection` (If we want to add more functionality with respect to feature selection)

Comment (neelasha23) [reply docs/_toc.yml:39]: Thoughts? @edublancas

File: docs/dimensionality_reduction/pca.md (hunk 4 above)
Comment (neelasha23) [reply docs/dimensionality_reduction/pca.md]: Have added the less data part. The time keeps changing with every run (sometimes is same,  sometimes less with PCA). So omitted it for now. It might be more obvious when we have a larger real-life dataset with lots of features but I don't think we should add it in the guide.

**Comment by neelasha23** [comment]
> I think we also need to add telemetry on this plot and test for it's mocked call.
> 

Telemetry call is added to the lot function.
However,  I don't see telemetry testcases in any of the plot test files.
There's a separate test file `test_skeval_telemetry` for the telemetry cases. Should I still add it in the pca test file? . Let me know what you think

> * I didn't see a changelog entry.

Added

@idomic

File: docs/_toc.yml (hunk 5 above)
Comment (edublancas) [reply docs/_toc.yml:39]: I think let's merge what we have, later we can think how to re-org the docs if needed.

**Comment by idomic** [comment]
@neelasha23 re: test_skeval_telemetry, add a test there worst case we move it to the pca module, we're missing some tests there but we can tackle it in a different issue.

File: docs/_toc.yml (hunk 5 above)
Comment (idomic) [reply docs/_toc.yml:39]: Agree, this has been going on for a while, it should happen soon due to SEO.

File: docs/dimensionality_reduction/pca.md (hunk 4 above)
Comment (idomic) [reply docs/dimensionality_reduction/pca.md]: Right, the point is to showcase the benefit, we're achieving the same result with less compute essentially.
You can resolve this.

**Comment by idomic** [comment]
Just finish it up (I think the last that's open is telemetry), fix conflict and we're good to go!

**Review by edublancas — APPROVED** [review]
